File: backend/src/routes/organizerRoutes.js
import express from 'express';
import { queueEmail, sendConfirmation } from '../services/emailService.js';
import { updateRegistration } from '../services/registrationService.js';

export const organizerRouter = express.Router();

organizerRouter.patch('/registrations/:id', async (request, response, next) => {
  try {
    const provided = request.headers.authorization?.replace(/^Bearer\s+/i, '');
    if (!process.env.ORGANIZER_SECRET || provided !== process.env.ORGANIZER_SECRET) return response.status(401).json({ message: 'Organizer authorization required.' });
    const allowed = ['confirmed', 'mismatch', 'duplicate'];
    if (!allowed.includes(request.body.status)) return response.status(400).json({ message: 'Invalid registration status.' });
    const update = { paymentStatus: request.body.status, verifiedAt: new Date(), verifiedBy: 'organizer-api', verificationNotes: String(request.body.notes || '').slice(0, 500) };
    const registration = await updateRegistration(request.params.id, update);
    if (!registration) return response.status(404).json({ message: 'Registration not found.' });
    if (update.paymentStatus === 'confirmed') queueEmail(() => sendConfirmation(registration));
    return response.json({ registrationId: registration.registrationId, status: registration.paymentStatus });
  } catch (error) {
    return next(error);
  }
});

File: backend/app/routes/organizer_routes.py
from datetime import datetime, timezone

from fastapi import APIRouter, Header, HTTPException, Request

from app.core.config import settings
from app.services.email_service import queue_email, send_confirmation
from app.services.registration_service import update_registration

router = APIRouter()


@router.patch("/registrations/{registration_id}")
async def legacy_verify(registration_id: str, request: Request, authorization: str | None = Header(default=None)):
    provided = str(authorization or "").replace("Bearer ", "", 1).replace("bearer ", "", 1)
    if not settings.organizer_secret or provided != settings.organizer_secret:
        raise HTTPException(status_code=401, detail="Organizer authorization required.")
    body = await request.json()
    if body.get("status") not in ["confirmed", "mismatch", "duplicate"]:
        raise HTTPException(status_code=400, detail="Invalid registration status.")
    update = {"paymentStatus": body["status"], "verifiedAt": datetime.now(timezone.utc), "verifiedBy": "organizer-api", "verificationNotes": str(body.get("notes") or "")[:500]}
    registration = await update_registration(registration_id, update)
    if not registration:
        raise HTTPException(status_code=404, detail="Registration not found.")
    if update["paymentStatus"] == "confirmed":
        queue_email(lambda: send_confirmation(registration))
    return {"registrationId": registration.get("registrationId"), "status": registration.get("paymentStatus")}
